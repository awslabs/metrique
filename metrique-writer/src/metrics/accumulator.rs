// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use std::{
    collections::HashMap,
    fmt::{Debug, Formatter},
    sync::{
        Arc, RwLock,
        atomic::{AtomicU64, Ordering},
    },
    time::SystemTime,
};

use super::unit::metrics_unit_to_metrique_unit;
use crate::metrics::metrics_histogram::Bucket;
use metrics::{Counter, Gauge, Histogram, Key, KeyName, Metadata, Recorder, SharedString, Unit};
use metrics_util::registry::{Registry, Storage};
use metrique_writer_core::{Entry, EntryWriter, Observation, value::MetricFlags};

/// A [`metrics_util::Storage`] that uses [`crate::metrics_histogram::Histogram`] for its histogram implementation.
pub struct AtomicStorageWithHistogram;

impl<K> Storage<K> for AtomicStorageWithHistogram {
    type Counter = Arc<AtomicU64>;
    type Gauge = Arc<AtomicU64>;
    type Histogram = Arc<crate::metrics::metrics_histogram::Histogram>;

    fn counter(&self, _: &K) -> Self::Counter {
        Arc::new(AtomicU64::new(0))
    }

    fn gauge(&self, _: &K) -> Self::Gauge {
        Arc::new(AtomicU64::new(0))
    }

    fn histogram(&self, _: &K) -> Self::Histogram {
        Arc::new(crate::metrics::metrics_histogram::Histogram::new())
    }
}

struct MetricRecorderInner {
    emit_zero_counters: bool,
    registry: Registry<metrics::Key, AtomicStorageWithHistogram>,
    units: RwLock<HashMap<String, metrique_writer_core::Unit>>,
}

/// The metric recorder belonging to this crate. Accumulates metrics in a registry
/// and lets them be read out via `readout`
#[derive(Clone)]
pub struct MetricRecorder(Arc<MetricRecorderInner>);

impl MetricRecorder {
    /// Create a new metric recorder
    pub fn new() -> Self {
        Self::new_with_emit_zero_counters(false)
    }

    /// Create a new metric recorder
    ///
    /// If `emit_zero_counters` is true, counters with a value of 0 will be emitted
    pub fn new_with_emit_zero_counters(emit_zero_counters: bool) -> Self {
        Self(Arc::new(MetricRecorderInner::new(emit_zero_counters)))
    }

    /// Read out the current value of the metrics, resetting counters and histograms (and
    /// not resetting gauges).
    pub fn readout(&self) -> MetricAccumulatorEntry {
        self.0.readout()
    }
}

impl Default for MetricRecorder {
    fn default() -> Self {
        Self::new()
    }
}

impl MetricRecorderInner {
    fn new(emit_zero_counters: bool) -> Self {
        Self {
            emit_zero_counters,
            registry: Registry::new(AtomicStorageWithHistogram),
            units: RwLock::new(HashMap::new()),
        }
    }

    fn readout(&self) -> MetricAccumulatorEntry {
        let mut counters = Vec::new();
        let mut gauges = Vec::new();
        let mut histograms = Vec::new();
        self.registry.visit_counters(|key, counter| {
            let counter = counter.swap(0, Ordering::Relaxed);
            // don't include counters that weren't incremented in the log
            if self.emit_zero_counters || counter != 0 {
                counters.push((key.clone(), counter));
            }
        });
        counters.sort_by(|u, v| u.0.cmp(&v.0));
        self.registry.visit_gauges(|key, gauge| {
            gauges.push((key.clone(), f64::from_bits(gauge.load(Ordering::Relaxed))));
        });
        gauges.sort_by(|u, v| u.0.cmp(&v.0));
        self.registry.visit_histograms(|key, histogram| {
            histograms.push((key.clone(), histogram.drain()));
        });
        histograms.sort_by(|u, v| u.0.cmp(&v.0));
        MetricAccumulatorEntry {
            counters,
            gauges,
            histograms,
            units: self.units.read().unwrap().clone(),
            timestamp: Some(SystemTime::now()),
        }
    }
}

/// Represents a readout of metrics, with values for all the given metrics.
#[derive(Clone, Debug)]
pub struct MetricAccumulatorEntry {
    counters: Vec<(metrics::Key, u64)>,
    gauges: Vec<(metrics::Key, f64)>,
    histograms: Vec<(metrics::Key, Vec<Bucket>)>,
    units: HashMap<String, metrique_writer_core::Unit>,
    timestamp: Option<SystemTime>,
}

impl MetricAccumulatorEntry {
    /// Remove the timestamp from this MetricAccumulatorEntry. This should be used
    /// if it is nested in a different metrics struct to avoid double timestamp
    /// recording, which might cause the metrics writer to panic.
    pub fn remove_timestamp(&mut self) {
        self.timestamp = None;
    }

    /// Get the current timestamp from this metrics accumulator entry
    pub fn timestamp(&self) -> Option<SystemTime> {
        self.timestamp
    }
}

#[cfg(any(test, feature = "test-util"))]
impl MetricAccumulatorEntry {
    /// Get counter value. O(n) in number of metrics so use only for tests.
    ///
    /// Use the `Entry` implementation when performance is needed.
    pub fn counter_value(&self, name: &str) -> Option<u64> {
        self.counters
            .iter()
            .find(|(key, _)| key.name() == name)
            .map(|(_, v)| *v)
    }

    /// Get gauge value. O(n) in number of metrics so use only for tests.
    ///
    /// Use the `Entry` implementation when performance is needed.
    pub fn gauge_value(&self, name: &str) -> Option<f64> {
        self.gauges
            .iter()
            .find(|(key, _)| key.name() == name)
            .map(|(_, v)| *v)
    }

    /// Get a list of histogram samples. O(n) in number of histograms so use only for tests
    ///
    /// Note that histograms use sampling which means that the result is somewhat inaccurate.
    ///
    /// Use the `Entry` implementation when performance is needed.
    pub fn histogram_value(&self, name: &str) -> Vec<u32> {
        self.histograms
            .iter()
            .filter(|(key, _)| key.name() == name)
            .flat_map(|(_key, buckets)| buckets)
            .flat_map(|bucket| vec![bucket.value; bucket.count as usize])
            .collect()
    }
}

impl Entry for MetricAccumulatorEntry {
    fn write<'a>(&'a self, writer: &mut impl EntryWriter<'a>) {
        struct MultiObservation<'a, T> {
            value: T,
            unit: metrique_writer_core::Unit,
            dimensions: Vec<(&'a str, &'a str)>,
        }

        impl<T> metrique_writer_core::Value for MultiObservation<'_, T>
        where
            T: IntoIterator<Item = Observation> + Clone,
        {
            fn write(&self, writer: impl metrique_writer_core::ValueWriter) {
                writer.metric(
                    self.value.clone(),
                    self.unit,
                    self.dimensions.iter().cloned(),
                    MetricFlags::empty(),
                )
            }
        }

        if let Some(timestamp) = self.timestamp {
            writer.timestamp(timestamp);
        }

        // Reporting time-based metrics, split entries is what we want.
        writer.config(&const { metrique_writer_core::config::AllowSplitEntries::new() });

        for (key, value) in &self.counters {
            let labels = key
                .labels()
                .map(|label| (label.key(), label.value()))
                .collect();
            let unit = self
                .units
                .get(key.name())
                .unwrap_or(&metrique_writer_core::Unit::None);
            writer.value(
                key.name(),
                &MultiObservation {
                    value: [Observation::Unsigned(*value)],
                    unit: *unit,
                    dimensions: labels,
                },
            );
        }

        for (key, value) in &self.gauges {
            let labels = key
                .labels()
                .map(|label| (label.key(), label.value()))
                .collect();
            let unit = self
                .units
                .get(key.name())
                .unwrap_or(&metrique_writer_core::Unit::None);
            writer.value(
                key.name(),
                &MultiObservation {
                    value: [Observation::Floating(*value)],
                    unit: *unit,
                    dimensions: labels,
                },
            );
        }

        for (key, buckets) in &self.histograms {
            let labels = key
                .labels()
                .map(|label| (label.key(), label.value()))
                .collect();
            let unit = self
                .units
                .get(key.name())
                .unwrap_or(&metrique_writer_core::Unit::None);
            let observations = buckets.iter().map(|bucket| Observation::Repeated {
                total: bucket.value as f64 * bucket.count as f64,
                occurrences: bucket.count as u64,
            });
            writer.value(
                key.name(),
                &MultiObservation {
                    value: observations,
                    unit: *unit,
                    dimensions: labels,
                },
            );
        }
    }
}

impl Recorder for MetricRecorder {
    fn describe_counter(
        &self,
        key: metrics::KeyName,
        unit: Option<metrics::Unit>,
        _description: metrics::SharedString,
    ) {
        self.0
            .units
            .write()
            .unwrap()
            .insert(key.as_str().to_string(), metrics_unit_to_metrique_unit(unit));
    }

    fn describe_gauge(
        &self,
        key: metrics::KeyName,
        unit: Option<metrics::Unit>,
        _description: metrics::SharedString,
    ) {
        self.0
            .units
            .write()
            .unwrap()
            .insert(key.as_str().to_string(), metrics_unit_to_metrique_unit(unit));
    }

    fn describe_histogram(
        &self,
        key: metrics::KeyName,
        unit: Option<metrics::Unit>,
        _description: metrics::SharedString,
    ) {
        self.0
            .units
            .write()
            .unwrap()
            .insert(key.as_str().to_string(), metrics_unit_to_metrique_unit(unit));
    }

    fn register_counter(
        &self,
        key: &metrics::Key,
        _metadata: &metrics::Metadata<'_>,
    ) -> metrics::Counter {
        metrics::Counter::from_arc(self.0.registry.get_or_create_counter(key, Clone::clone))
    }

    fn register_gauge(
        &self,
        key: &metrics::Key,
        _metadata: &metrics::Metadata<'_>,
    ) -> metrics::Gauge {
        metrics::Gauge::from_arc(self.0.registry.get_or_create_gauge(key, Clone::clone))
    }

    fn register_histogram(
        &self,
        key: &metrics::Key,
        _metadata: &metrics::Metadata<'_>,
    ) -> metrics::Histogram {
        metrics::Histogram::from_arc(self.0.registry.get_or_create_histogram(key, Clone::clone))
    }
}

/// A Cloneable dynamic recorder that implements the Recorder trait
#[derive(Clone)]
pub struct SharedRecorder(Arc<dyn Recorder + Send + Sync>);
impl SharedRecorder {
    pub fn new(recorder: Arc<dyn Recorder + Send + Sync>) -> Self {
        Self(recorder)
    }
}

impl Debug for SharedRecorder {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("SharedRecorder").finish()
    }
}

impl Recorder for SharedRecorder {
    fn describe_counter(&self, key: KeyName, unit: Option<Unit>, description: SharedString) {
        self.0.describe_counter(key, unit, description);
    }

    fn describe_gauge(&self, key: KeyName, unit: Option<Unit>, description: SharedString) {
        self.0.describe_gauge(key, unit, description);
    }

    fn describe_histogram(&self, key: KeyName, unit: Option<Unit>, description: SharedString) {
        self.0.describe_histogram(key, unit, description);
    }

    fn register_counter(&self, key: &Key, metadata: &Metadata<'_>) -> Counter {
        self.0.register_counter(key, metadata)
    }

    fn register_gauge(&self, key: &Key, metadata: &Metadata<'_>) -> Gauge {
        self.0.register_gauge(key, metadata)
    }

    fn register_histogram(&self, key: &Key, metadata: &Metadata<'_>) -> Histogram {
        self.0.register_histogram(key, metadata)
    }
}

#[cfg(test)]
mod test {
    use metrics::{histogram, with_local_recorder};
    use metrique_writer_core::{format::Format, test_stream::DummyFormat};
    use test_case::test_case;

    use crate::metrics::MetricRecorder;

    #[test_case(false, None; "no_emit_zero_counters")]
    #[test_case(true, Some(0); "emit_zero_counters")]
    fn test_emit_zero_counters(emit_zero_counters: bool, expected_result: Option<u64>) {
        let accumulator: MetricRecorder =
            MetricRecorder::new_with_emit_zero_counters(emit_zero_counters);
        metrics::with_local_recorder(&accumulator, || {
            metrics::counter!("a").increment(1);
        });
        let read0 = accumulator.readout();
        assert_eq!(read0.counter_value("a"), Some(1));
        let read1 = accumulator.readout();
        assert_eq!(read1.counter_value("a"), expected_result);
    }

    #[test]
    fn simple() {
        let recorder = MetricRecorder::new();
        with_local_recorder(&recorder, || {
            let histogram = histogram!("test");
            histogram.record(1);
            histogram.record(2);
            histogram.record(2);
            histogram.record(3);
            histogram.record(3);
            histogram.record(3);
        });
        let mut readout = recorder.readout();
        // force some timestamp for test purposes
        readout.timestamp =
            Some(std::time::SystemTime::UNIX_EPOCH + std::time::Duration::from_secs(86_400));
        let mut writer = DummyFormat;
        let mut output = Vec::new();

        writer.format(&readout, &mut output).unwrap();

        let output = String::from_utf8(output).unwrap();
        assert_eq!(
            output,
            r#"[("timestamp", "86400"), ("test", "[Repeated { total: 1.0, occurrences: 1 }, Repeated { total: 4.0, occurrences: 2 }, Repeated { total: 9.0, occurrences: 3 }] None []")]"#
        );
    }
}
