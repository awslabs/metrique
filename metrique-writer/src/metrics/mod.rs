// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

//! This mode providesa few [`metrics::Recorder`]s that can be used for emitting metrics
//! via metrique-writer. This includes [`MetricsReporter`]  that is designed for use in EC2/Fargate,
//! [`lambda_reporter`] that is designed for use in Lambda, and [`capture`] that is
//! designed for use in unit tests.
//!
//! [`metrics::Recorder`]: metrics::Recorder
//! [`MetricsReporter`]: crate::metrics::MetricReporter
//! [`lambda_reporter`]: crate::metrics::lambda_reporter
//! [`capture`]: crate::metrics::capture

pub(crate) mod accumulator;
pub mod capture;
pub mod lambda_reporter;
pub mod metrics_histogram;
mod reporter;
mod unit;

pub use accumulator::{MetricAccumulatorEntry, MetricRecorder, SharedRecorder};
pub use reporter::{MetricReporter, MetricReporterBuilder};
