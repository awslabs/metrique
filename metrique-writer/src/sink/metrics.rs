// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

/// Defines callbacks for recording metrics
pub trait MetricRecorder {
    /// Records a histogram entry. metric is used to define the metric, sink to define the sink, value the histogram value
    fn record_histogram(&self, metric: &'static str, sink: &str, value: u32);
    /// Increments a histogram entry. metric is used to define the metric, sink to define the sink, value the histogram value
    fn increment_counter(&self, metric: &'static str, sink: &str, value: u64);
    /// Sets a gauge entry. metric is used to define the metric, sink to define the sink, value the histogram value
    fn set_gauge(&self, metric: &'static str, sink: &str, value: f64);
    /// Increments a gauge entry. metric is used to define the metric, sink to define the sink, value the histogram value
    fn increment_gauge(&self, metric: &'static str, sink: &str, value: f64);
    /// Decrements a gauge entry. metric is used to define the metric, sink to define the sink, value the histogram value
    fn decrement_gauge(&self, metric: &'static str, sink: &str, value: f64);
}

/// Implements MetricRecorder for a global metrics-rs 0.24 recorder
#[cfg(feature = "metrics_rs_024")]
#[derive(Debug, Copy, Clone)]
pub(crate) struct GlobalMetricsRs024Bridge;

#[cfg(feature = "metrics_rs_024")]
impl MetricRecorder for GlobalMetricsRs024Bridge {
    fn record_histogram(&self, metric: &'static str, sink: &str, value: u32) {
        metrics::histogram!(metric, "sink" => sink.to_owned()).record(value);
    }

    fn increment_counter(&self, metric: &'static str, sink: &str, value: u64) {
        metrics::counter!(metric, "sink" => sink.to_owned()).increment(value);
    }

    fn set_gauge(&self, metric: &'static str, sink: &str, value: f64) {
        metrics::gauge!(metric, "sink" => sink.to_owned()).set(value);
    }

    fn increment_gauge(&self, metric: &'static str, sink: &str, value: f64) {
        metrics::gauge!(metric, "sink" => sink.to_owned()).increment(value);
    }

    fn decrement_gauge(&self, metric: &'static str, sink: &str, value: f64) {
        metrics::gauge!(metric, "sink" => sink.to_owned()).decrement(value);
    }
}

/// Implements MetricRecorder for a local metrics-rs 0.24 recorder
#[cfg(feature = "metrics_rs_024")]
#[derive(Debug, Copy, Clone)]
pub(crate) struct LocalMetricsRs024Bridge<R>(pub(crate) R);

#[cfg(feature = "metrics_rs_024")]
impl<R: metrics::Recorder> MetricRecorder for LocalMetricsRs024Bridge<R> {
    fn record_histogram(&self, metric: &'static str, sink: &str, value: u32) {
        self.0
            .register_histogram(
                &metrics::Key::from_parts(
                    metric,
                    vec![metrics::Label::new("sink", sink.to_owned())],
                ),
                &metrics::Metadata::new(module_path!(), metrics::Level::INFO, Some(module_path!())),
            )
            .record(value);
    }

    fn increment_counter(&self, metric: &'static str, sink: &str, value: u64) {
        self.0
            .register_counter(
                &metrics::Key::from_parts(
                    metric,
                    vec![metrics::Label::new("sink", sink.to_owned())],
                ),
                &metrics::Metadata::new(module_path!(), metrics::Level::INFO, Some(module_path!())),
            )
            .increment(value);
    }

    fn set_gauge(&self, metric: &'static str, sink: &str, value: f64) {
        self.0
            .register_gauge(
                &metrics::Key::from_parts(
                    metric,
                    vec![metrics::Label::new("sink", sink.to_owned())],
                ),
                &metrics::Metadata::new(module_path!(), metrics::Level::INFO, Some(module_path!())),
            )
            .set(value);
    }

    fn increment_gauge(&self, metric: &'static str, sink: &str, value: f64) {
        self.0
            .register_gauge(
                &metrics::Key::from_parts(
                    metric,
                    vec![metrics::Label::new("sink", sink.to_owned())],
                ),
                &metrics::Metadata::new(module_path!(), metrics::Level::INFO, Some(module_path!())),
            )
            .increment(value);
    }

    fn decrement_gauge(&self, metric: &'static str, sink: &str, value: f64) {
        self.0
            .register_gauge(
                &metrics::Key::from_parts(
                    metric,
                    vec![metrics::Label::new("sink", sink.to_owned())],
                ),
                &metrics::Metadata::new(module_path!(), metrics::Level::INFO, Some(module_path!())),
            )
            .decrement(value);
    }
}

pub(crate) trait GlobalRecorderVersion {
    fn recorder() -> impl MetricRecorder + Send + 'static;
    fn describe(metrics: &[DescribedMetric]);
}

#[cfg(feature = "metrics_rs_024")]
impl GlobalRecorderVersion for dyn metrics::Recorder {
    fn describe(metrics: &[DescribedMetric]) {
        for metric in metrics {
            let unit = match metric.unit {
                MetricsRsUnit::Count => metrics::Unit::Count,
                MetricsRsUnit::Percent => metrics::Unit::Percent,
                MetricsRsUnit::Millisecond => metrics::Unit::Milliseconds,
            };
            match metric.r#type {
                MetricsRsType::Counter => {
                    metrics::describe_counter!(metric.name, unit, metric.description)
                }
                MetricsRsType::Gauge => {
                    metrics::describe_gauge!(metric.name, unit, metric.description)
                }
                MetricsRsType::Histogram => {
                    metrics::describe_histogram!(metric.name, unit, metric.description)
                }
            }
        }
    }

    fn recorder() -> impl MetricRecorder {
        GlobalMetricsRs024Bridge
    }
}

/// Describes a metrics.rs unit in a non-exhaustive fashion
#[non_exhaustive]
#[derive(Copy, Clone, Debug)]
pub enum MetricsRsUnit {
    Percent,
    Count,
    Millisecond,
}

/// Describes a metrics.rs metric type in a non-exhaustive fashion
#[non_exhaustive]
#[derive(Copy, Clone, Debug)]
pub enum MetricsRsType {
    Gauge,
    Counter,
    Histogram,
}

#[derive(Copy, Clone, Debug)]
#[non_exhaustive]
pub struct DescribedMetric {
    pub name: &'static str,
    pub unit: MetricsRsUnit,
    pub r#type: MetricsRsType,
    pub description: &'static str,
}

pub trait LocalRecorderVersion<R> {
    fn recorder(recorder: R) -> impl MetricRecorder + Send + 'static;
}

#[cfg(feature = "metrics_rs_024")]
impl<R> LocalRecorderVersion<R> for dyn metrics::Recorder
where
    R: metrics::Recorder + Send + 'static,
{
    fn recorder(recorder: R) -> impl MetricRecorder + Send + 'static {
        LocalMetricsRs024Bridge(recorder)
    }
}
