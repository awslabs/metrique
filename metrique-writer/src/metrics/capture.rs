// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

//! This module contains functions for capturing metrics, generally to be used in tests.

use std::future::Future;

use metrics::Recorder;
use pin_project::pin_project;

use crate::metrics::{MetricAccumulatorEntry, MetricRecorder};

/// Run `f`, capturing the metrics while it runs using a local recorder.
pub fn capture_metrics<T, F: FnOnce() -> T>(f: F) -> (MetricAccumulatorEntry, T) {
    let accumulator: MetricRecorder = MetricRecorder::new();
    let res = metrics::with_local_recorder(&accumulator, f);
    (accumulator.readout(), res)
}

/// Asynchrounously run `f`, capturing the metrics while it runs using a local recorder.
///
/// If `f` spawns subtasks, metrics from the subtasks will *not* be captured.
pub async fn capture_metrics_async<T, F: Future<Output = T>>(f: F) -> (MetricAccumulatorEntry, T) {
    let accumulator = MetricRecorder::new();
    let res = LocalRecorderWrapper::new(accumulator.clone(), f).await;
    (accumulator.readout(), res)
}

/// Wraps a future to install a local recorder during the executor of said future.
#[pin_project]
pub struct LocalRecorderWrapper<R: Recorder, F: Future> {
    recorder: R,
    #[pin]
    future: F,
}

impl<R: Recorder, F: Future> Future for LocalRecorderWrapper<R, F> {
    type Output = F::Output;

    fn poll(
        self: std::pin::Pin<&mut Self>,
        cx: &mut std::task::Context<'_>,
    ) -> std::task::Poll<Self::Output> {
        let this = self.project();
        metrics::with_local_recorder(this.recorder, || this.future.poll(cx))
    }
}

impl<R: Recorder, F: Future> LocalRecorderWrapper<R, F> {
    /// Create a new `LocalRecorderWrapper`
    pub fn new(recorder: R, future: F) -> Self {
        Self { recorder, future }
    }
}

#[cfg(test)]
mod test {
    #[test]
    fn test_capture_metrics() {
        let (metrics, _) = super::capture_metrics(|| {
            metrics::counter!("foo").increment(9);
            metrics::counter!("foo").increment(3);
            metrics::gauge!("bar").set(5);
            metrics::histogram!("baz").record(100);
            metrics::histogram!("baz").record(101);
            metrics::histogram!("baz").record(1000);
        });
        assert_eq!(metrics.counter_value("foo"), Some(12));
        assert_eq!(metrics.counter_value("nothing"), None);
        assert_eq!(metrics.gauge_value("bar"), Some(5.0));
        assert_eq!(metrics.gauge_value("nothing"), None);
        // FIXME: use fuzzy matching for histogram?
        assert_eq!(metrics.histogram_value("baz"), vec![101, 101, 1007]);
        assert_eq!(metrics.histogram_value("nothing"), Vec::<u32>::new());
    }

    #[tokio::test]
    async fn test_capture_metrics_async() {
        let (metrics, _) = super::capture_metrics_async(async move {
            metrics::counter!("foo").increment(9);
            metrics::counter!("foo").increment(3);
            tokio::task::yield_now().await;
            metrics::gauge!("bar").set(5);
            metrics::histogram!("baz").record(100);
            metrics::histogram!("baz").record(101);
            metrics::histogram!("baz").record(1000);
        })
        .await;
        assert_eq!(metrics.counter_value("foo"), Some(12));
        assert_eq!(metrics.counter_value("nothing"), None);
        assert_eq!(metrics.gauge_value("bar"), Some(5.0));
        assert_eq!(metrics.gauge_value("nothing"), None);
        // FIXME: use fuzzy matching for histogram?
        assert_eq!(metrics.histogram_value("baz"), vec![101, 101, 1007]);
        assert_eq!(metrics.histogram_value("nothing"), Vec::<u32>::new());
    }
}
