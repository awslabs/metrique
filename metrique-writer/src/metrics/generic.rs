use std::{collections::HashMap, hash::Hash, sync::atomic::Ordering, time::SystemTime};

use crate::metrics::{
    MetricAccumulatorEntry, MetricRecorder, accumulator::AtomicStorageWithHistogram,
};

mod private {
    pub trait Sealed {}
    #[cfg(feature = "metrics_rs_024")]
    impl Sealed for dyn metrics::Recorder {}

    pub trait Sealed2<V: super::MetricsRsVersion + ?Sized> {}
}

/// A trait to allow the metrics.rs bridge to be generic over metrics.rs versions
///
/// This is not to be implemented or called directly by users.
pub trait MetricsRsVersion: 'static + private::Sealed {
    #[doc(hidden)]
    type Key: Hash + Send;
    #[doc(hidden)]
    type AtomicStorageWithHistogramRegistry: Send + Sync;
    #[doc(hidden)]
    type Recorder: ?Sized;
    #[doc(hidden)]
    fn new_atomic_storage_with_histogram_registry() -> Self::AtomicStorageWithHistogramRegistry;
    #[doc(hidden)]
    fn readout(
        registry: &Self::AtomicStorageWithHistogramRegistry,
        emit_zero_counters: bool,
        units: impl FnOnce() -> HashMap<String, metrique_writer_core::Unit>,
    ) -> MetricAccumulatorEntry<Self>;
    #[doc(hidden)]
    fn key_name(name: &Self::Key) -> &str;
    #[doc(hidden)]
    fn key_labels(key: &Self::Key) -> Vec<(&str, &str)>;
    #[doc(hidden)]
    fn set_global_recorder(recorder: MetricRecorder<Self>);
}

#[cfg(feature = "metrics_rs_024")]
impl MetricsRsVersion for dyn metrics::Recorder {
    #[doc(hidden)]
    type Key = metrics::Key;
    #[doc(hidden)]
    type AtomicStorageWithHistogramRegistry =
        metrics_util::registry::Registry<metrics::Key, AtomicStorageWithHistogram>;
    #[doc(hidden)]
    type Recorder = dyn metrics::Recorder + Send + Sync;
    #[doc(hidden)]
    fn new_atomic_storage_with_histogram_registry() -> Self::AtomicStorageWithHistogramRegistry {
        metrics_util::registry::Registry::new(AtomicStorageWithHistogram)
    }
    fn readout(
        registry: &Self::AtomicStorageWithHistogramRegistry,
        emit_zero_counters: bool,
        units: impl FnOnce() -> HashMap<String, metrique_writer_core::Unit>,
    ) -> MetricAccumulatorEntry<Self> {
        let mut counters = Vec::new();
        let mut gauges = Vec::new();
        let mut histograms = Vec::new();
        registry.visit_counters(|key, counter| {
            let counter = counter.swap(0, Ordering::Relaxed);
            // don't include counters that weren't incremented in the log
            if emit_zero_counters || counter != 0 {
                counters.push((key.clone(), counter));
            }
        });
        counters.sort_by(|u, v| u.0.cmp(&v.0));
        registry.visit_gauges(|key, gauge| {
            gauges.push((key.clone(), f64::from_bits(gauge.load(Ordering::Relaxed))));
        });
        gauges.sort_by(|u, v| u.0.cmp(&v.0));
        registry.visit_histograms(|key, histogram| {
            histograms.push((key.clone(), histogram.drain()));
        });
        histograms.sort_by(|u, v| u.0.cmp(&v.0));
        MetricAccumulatorEntry {
            counters,
            gauges,
            histograms,
            units: units(),
            timestamp: Some(SystemTime::now()),
        }
    }
    fn key_name(key: &Self::Key) -> &str {
        key.name()
    }
    fn key_labels(key: &Self::Key) -> Vec<(&str, &str)> {
        key.labels()
            .map(|label| (label.key(), label.value()))
            .collect()
    }
    #[track_caller]
    fn set_global_recorder(recorder: MetricRecorder<Self>) {
        metrics::set_global_recorder(recorder).expect("failed to set global recorder");
    }
}

/// A trait to allow the metrics.rs bridge to be generic over versions of metrics.rs
/// [metrics::Recorder]s. This trait is not to be manually implemented or called by
/// users.
pub trait ParametricRecorder<V: MetricsRsVersion + ?Sized>: private::Sealed2<V> {
    #[doc(hidden)]
    fn with_local_recorder<T>(&self, body: impl FnOnce() -> T) -> T;
}

#[cfg(feature = "metrics_rs_024")]
impl<R: metrics::Recorder> ParametricRecorder<dyn metrics::Recorder> for R {
    fn with_local_recorder<T>(&self, body: impl FnOnce() -> T) -> T {
        metrics::with_local_recorder(self, body)
    }
}
#[cfg(feature = "metrics_rs_024")]
impl<R: metrics::Recorder> private::Sealed2<dyn metrics::Recorder> for R {}
