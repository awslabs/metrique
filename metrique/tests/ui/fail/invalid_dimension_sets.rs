// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use metrique::unit_of_work::metrics;

#[metrics(
    emf::dimension_sets = ["a", "b", "c"],
    rename_all = "snake_case"
)]
#[derive(Default, Clone)]
struct Nested {
    a: usize,
}

fn main() {}
