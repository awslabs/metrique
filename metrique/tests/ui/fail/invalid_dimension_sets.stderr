error: expected a list of dimensions like `["Service", "Operation"]`
 --> tests/ui/fail/invalid_dimension_sets.rs:7:28
  |
7 |     emf::dimension_sets = ["a", "b", "c"],
  |                            ^^^

error: expected a list of dimensions like `["Service", "Operation"]`
 --> tests/ui/fail/invalid_dimension_sets.rs:7:33
  |
7 |     emf::dimension_sets = ["a", "b", "c"],
  |                                 ^^^

error: expected a list of dimensions like `["Service", "Operation"]`
 --> tests/ui/fail/invalid_dimension_sets.rs:7:38
  |
7 |     emf::dimension_sets = ["a", "b", "c"],
  |                                      ^^^
